### PR TITLE
Regions support

### PR DESCRIPTION
**Pull request by ksitak**

### CODE REVIEW

File: config/config_test.go [hunk 1]
```diff
@@ -347,9 +305,20 @@ func TestShouldNotPassValidateConfigurationHTTPHandlerWithWrongRequests(t *testi
...
+	result, _ := ValidateConf(testConfData, true)
+
+	assert.False(t, result)
+}
+
 func PrepareYamlConfig(bodyMaxSize shardingconfig.HumanSizeUnits, idleConnTimeoutInp time.Duration, responseHeaderTimeoutInp time.Duration,
-	maintainedBackendHost string, listen string, technicalEndpointListen string, clientCfgName string,
-	clientClusters []string) YamlConfig {
+	maintainedBackendHost string, listen string, technicalEndpointListen string,
+	regions map[string]shardingconfig.RegionConfig) YamlConfig {
```
Comment (janisz) [inline config/config_test.go]: Too many parameters! Move them to structure.

File: config/validator.go [hunk 2]
```diff
@@ -60,25 +60,38 @@ func UniqueValuesInSliceValidator(v interface{}, param string) error {
...
 
-// ClientClustersEntryLogicalValidator validate Client->Clusters entry and make sure that all required clusters are defined
-func (c *YamlConfig) ClientClustersEntryLogicalValidator(valid *bool, validationErrors *map[string][]error) {
-	errorsList := make(map[string][]error)
-
-	if len(c.Client.Clusters) == 0 {
-		*valid = false
-		errorDetail := []error{errors.New("Empty clusters definition")}
-		errorsList["ClientClustersEntryLogicalValidator"] = errorDetail
+//RegionsEntryLogicalValidator validator for Regions part
```
Comment (janisz) [inline config/validator.go]: This does not tell anything. Please describe what/why it's doing

File: config/validator.go [hunk 3]
```diff
@@ -60,25 +60,38 @@ func UniqueValuesInSliceValidator(v interface{}, param string) error {
...
+func (c *YamlConfig) RegionsEntryLogicalValidator(valid *bool, validationErrors *map[string][]error) {
```
Comment (janisz) [inline config/validator.go]: output parameters? This should be return of this function.

File: sharding/config/config.go [hunk 4]
```diff
@@ -13,21 +13,23 @@ import (
...
+	Cluster string `yaml:"Cluster"`
+	//Cluster weight
+	Weight int `yaml:"Weight"`
+}
+
+//RegionConfig region configuration
+type RegionConfig struct {
+	//Multi cluster config
+	Clusters []MultiClusterConfig `yaml:"Clusters"`
+	//domains used for region matching
```
Comment (janisz) [inline sharding/config/config.go]: `d` → `D`

File: sharding/sharding.go [hunk 5]
```diff
@@ -1,85 +1,28 @@
...
-	if ok {
-		return s3cluster, nil
-	}
-	s3cluster, err := rf.initCluster(name)
-	if err != nil {
-		return s3cluster, err
-	}
-	rf.clusters[name] = s3cluster
-	return s3cluster, nil
+	storages  *storages.Storages
```
Comment (janisz) [inline sharding/sharding.go]: why using pointer?

File: sharding/shards_ring.go [hunk 6]
```diff
@@ -133,8 +139,8 @@ func (sr *ShardsRing) logInconsistency(key, expectedClusterName, actualClusterNa
 	}
 }
 
-// RoundTrip performs http requests
-func (sr ShardsRing) RoundTrip(req *http.Request) (resp *http.Response, rerr error) {
+//DoRequest performs http requests
```
Comment (janisz) [inline sharding/shards_ring.go]: What it does? What it returns and when? How it handle nil? Compare with https://golang.org/pkg/net/http/#Client.Do

File: storages/storages.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+func (st Storages) initCluster(name string) (Cluster, error) {
+	clusterConf, ok := st.Conf.Clusters[name]
+	if !ok {
+		return Cluster{}, fmt.Errorf("no cluster %q in configuration", name)
+	}
+	respHandler := httphandler.EarliestResponseHandler(st.Conf)
+	return newMultiBackendCluster(st.Transport, respHandler, clusterConf, name, st.Conf.MaintainedBackends), nil
+}
+
+//GetCluster gets cluster by name
```
Comment (janisz) [inline storages/storages.go]: define what it returns and when error is not nil

File: config/config_test.go (hunk 1 above)
Comment (ksitak) [inline config/config_test.go]: It's now one less than before!

File: config/validator.go (hunk 2 above)
Comment (ksitak) [inline config/validator.go]: Really? It validates "regions" part of configuration file... and tells if it's ok or not...

**Pull request merged**